argh_derive/src/help_json.rs

```diff
@@ -0,0 +1,234 @@
+            "{{\"name\": \"{}\", \"description\": \"{}\"}}",
+            pos.name,
+            escape_json(&pos.description)
+        ));
+    }
+    retval
+}
+
+/// Returns a `TokenStream` generating a `String` help message containing JSON.
+///
+/// Note: `fields` entries with `is_subcommand.is_some()` will be ignored
+/// in favor of the `subcommand` argument.
+pub(crate) fn help_json(
+    errors: &Errors,
+    cmd_name_str_array_ident: &syn::Ident,
+    ty_attrs: &TypeAttrs,
+    fields: &[StructField<'_>],
+    subcommand: Option<&StructField<'_>>,
+) -> TokenStream {
+    let mut usage_format_pattern = "{command_name}".to_string();
+    build_usage_command_line(&mut usage_format_pattern, fields, subcommand);
+
+    let mut help_obj = HelpJSON {
+        usage: String::from(""),
+        description: String::from(""),
+        positional_args: vec![],
+        options: vec![],
+        examples: String::from(""),
+        notes: String::from(""),
+        error_codes: vec![],
+    };
+
+    // Add positional args to the help object.
+    let positional = fields.iter().filter(|f| f.kind == FieldKind::Positional);
+    for arg in positional {
+        let mut description = String::from("");
+        if let Some(desc) = &arg.attrs.description {
+            description = desc.content.value().trim().to_owned();
+        }
+        help_obj.positional_args.push(PositionalHelp { name: arg.arg_name(), description });
+    }
+
+    // Add options to the help object.
+    let options = fields.iter().filter(|f| f.long_name.is_some());
+    for option in options {
+        let short = match option.attrs.short.as_ref().map(|s| s.value()) {
+            Some(c) => String::from(c),
+            None => String::from(""),
+        };
+        let long_with_leading_dashes =
+            option.long_name.as_ref().expect("missing long name for option");
+        let description =
+            require_description(errors, option.name.span(), &option.attrs.description, "field");
+        help_obj.options.push(OptionHelp {
+            short,
+            long: long_with_leading_dashes.to_owned(),
+            description,
+        });
+    }
+    // Also include "help" and "help-json"
+    help_obj.options.push(OptionHelp {
+        short: String::from(""),
+        long: String::from(HELP_FLAG),
+        description: String::from(HELP_DESCRIPTION),
+    });
+    help_obj.options.push(OptionHelp {
+        short: String::from(""),
+        long: String::from(HELP_JSON_FLAG),
+        description: String::from(HELP_JSON_DESCRIPTION),
+    });
+
+    let subcommand_calculation;
+    if let Some(subcommand) = subcommand {
+        let subcommand_ty = subcommand.ty_without_wrapper;
+        subcommand_calculation = quote! {
+            let mut subcommands = String::from("");
+            for cmd in  <#subcommand_ty as argh::SubCommands>::COMMANDS {
+                if !subcommands.is_empty() {
+                    subcommands.push_str(",\n    ");
```

Comment:
Do you think it would be better to omit the additional whitespace (newlines and spaces, here and below), except perhaps after colons and list commas perhaps? (I don't remember if, for example, python or chrome debugger prints spaces after colons and commas... maybe check on what's de facto standard for this, but I'm pretty sure they omit newlines and indentation.) The raw JSON output is really what you're looking for, I think, and it can always be pretty-printed. Otherwise, argh users have `--help` for more readable output.

Reply:
i removed the spaces, but kept the newlines between elements. The newlines make it much easier to debug the tests vs. a super long line.

Reply:
OK, I can see that being helpful when reviewing CI logs. SGTM